### PR TITLE
Remove Books from New Bookshelf Workflow

### LINKED ISSUE #295: Orcid workflow exception
Tasks - 

Send Fast Email
Send Student Email

Exception message:
org.camunda.bpm.engine.impl.pvm.PvmException: couldn't execute activity <serviceTask id="email_task_dcf2e6bb_7726_4fbd_ad39_f9009372beb0" ...>: The following has evaluated to null or missing:
==> emailTo  [in template "mailTo" at line 1, column 3]

### PR DESCRIPTION
**Pull request by wwelling**
Blocked by https://github.com/TAMULib/mod-camunda/pull/241

This workflow was carried over from Sprint16 to 17 because, in Sprint16 we prioritized QA. 

[sprint16-327-remove-books-nbs](https://github.com/TAMULib/fw-registry/commits/sprint17-327-remove-books-nbs/) -  This branch was cherry picked from this commit -  [b2e1074e](https://github.com/TAMULib/fw-registry/commit/b2e1074e05a9b8fc3d89db04e34d6b348cbd528a) and formed the basis of this workflow. This branch had merge conflicts with the main. 

sprint16-327-remove-books-nbs-merge - This commit was the branch resolution of the commit [b2e1074e](https://github.com/TAMULib/fw-registry/commit/b2e1074e05a9b8fc3d89db04e34d6b348cbd528a) but the transition from last sprint to this sprint was not smooth. so we have created new branches for sprint 17 below.

327-remove-books-from-nbs-by-barcodes - [b2e1074e](https://github.com/TAMULib/fw-registry/commit/b2e1074e05a9b8fc3d89db04e34d6b348cbd528a) - This commit is from sprint 16 and was not completed. It included partially done work on the workflow - remove books from nbs.

sprint17-327-remove-books-nbs
sprint17-staging
sprint17

Resolves #295

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -301,7 +312,7 @@ curl --location --request POST 'http://localhost:9001/mod-workflow/events/workfl
 --header 'X-Okapi-Tenant: tamu' \
 --form 'logLevel="DEBUG"' \
 --form 'file=@"FOLIOTags.csv"' \
---form 'path="/mnt/workflows/${tenantId}/create-tags"' \
+--form 'path="/mnt/workflows/diku/create-tags"' \
```
Comment (wwelling) [inline README.md:315]: https://github.com/TAMULib/fw-registry/pull/342#discussion_r1534165064

File: remove-books-from-nbs/nodes/email.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    },
+    {
+      "key": "email",
+      "type": "LOCAL",
+      "spin": true
+    }
+  ],
+  "outputVariable": {},
+  "mailTo": "${emailTo}",
+  "mailFrom": "${{nbs-mail-from}}",
```
Comment (kaladay) [inline remove-books-from-nbs/nodes/email.json:19]: Please fix this and replace `${{nbs-mail-from}}` with `${nbsMailFrom}`.
The `-` cannot be used in the names and we should only have a single braces not double braces.

File: README.md [hunk 3]
```diff
@@ -18,6 +18,17 @@ mvn clean spring-boot:run
 
 - Be sure to check and update the tenant header in all the curl requests documented below.
 
+## Variable substitution
```
[the remark this answers is not in the document]
Comment (kaladay) [reply README.md:21]: I didn't request for this to be added here.

Do we really want this documentation here?
I'm fine if it is, I just wanted to be clear that this was not the request I made.

My request was to either update the wiki if `${{}}` is valid or fix the invalid syntax.

Also, it is not freemarker syntax that we area dealing with.
It is camunda syntax.
see: https://docs.camunda.org/manual/7.18/user-guide/process-engine/expression-language/

**Review by kaladay — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (wwelling) [reply README.md:21]: Could you please provide a link to the Wiki in your request to update it?

[the remark this answers is not in the document]
Comment (wwelling) [reply README.md:21]: Oh and we didn't even correct it yet. Please see latest commit.

File: remove-books-from-nbs/nodes/email.json (hunk 2 above)
Comment (wwelling) [reply remove-books-from-nbs/nodes/email.json]: It should be `{{nbs-mail-from}}` as it is not a catalog reports form input variable. Do we want catalog reports defining these "global" to all workflows email from? 

Or are you requesting it be hardcoded in the catalog reports trigger form submit? Which issues POST request to mod-workflow.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (wwelling) [reply README.md:21]: Also, not all using `${}` syntax are processed by camunda. But I did miss those in the added description. 

However, the email task to email delegate uses freemarker in the delegate. Perhaps some improvements can be made.

https://github.com/TAMULib/mod-camunda/blob/main/src/main/java/org/folio/rest/delegate/EmailDelegate.java#L38
https://github.com/TAMULib/mod-camunda/blob/main/src/main/java/org/folio/rest/delegate/EmailDelegate.java#L70

[the remark this answers is not in the document]
Comment (wwelling) [reply README.md:21]: Here is an example of Camunda expression evaluation.

https://github.com/TAMULib/fw-registry/pull/343/files#diff-7d2047632ca971bdfd3275170d756fe41a0244395fd1bdd7572aa05d31a4fdb0R14

[the remark this answers is not in the document]
Comment (wwelling) [reply README.md:21]: Please see latest commit updating the readme.

**Comment by wwelling** [comment]
> My request was to either update the wiki if ${{}} is valid or fix the invalid syntax.

@kaladay the syntax `${{}}` is invalid I think.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (wwelling) [reply README.md:21]: > I didn't request for this to be added here.
> 
> Do we really want this documentation here? I'm fine if it is, I just wanted to be clear that this was not the request I made.
> 
> My request was to either update the wiki if `${{}}` is valid or fix the invalid syntax.
> 
> Also, it is not freemarker syntax that we area dealing with. It is camunda syntax. see: https://docs.camunda.org/manual/7.18/user-guide/process-engine/expression-language/

The change request comment is easily misinterpreted. Anyhow, I would add this documentation in the README.

image

**Review by Dbreck-TAMU — APPROVED** [review]

**Review by kaladay — CHANGES_REQUESTED** [review]
This set is to go into sprint17-staging.
I think this is the wrong staging branch.

File: remove-books-from-nbs/nodes/email.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      "key": "tenantId",
+      "type": "PROCESS"
+    }
+  ],
+  "outputVariable": {},
+  "mailTo": "${emailTo}",
+  "mailFrom": "{{nbs-mail-from}}",
+  "mailSubject": "${email.subject}",
+  "mailText": "${email.text}",
+  "mailMarkup": "${email.markup}",
```
Comment (kaladay) [inline remove-books-from-nbs/nodes/email.json:30]: This is inconsistent.
`emailTo` and `email.text`.

**Pull request merged**